examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp

```diff
@@ -14,95 +14,126 @@
+  // have to cast away const because MPI requires it
+  autopas::AutoPas_MPI_Cart_create(AUTOPAS_MPI_COMM_WORLD, _dimensionCount, _decomposition.data(),
+                                   const_cast<int *>(_periods.data()), true, &_communicator);
   autopas::AutoPas_MPI_Comm_rank(_communicator, &_domainIndex);
 }
 
 void RegularGridDecomposition::initializeLocalDomain() {
-  _domainId = {0, 0, 0};
-  autopas::AutoPas_MPI_Comm_rank(_communicator, &_domainIndex);
-
-  std::vector<int> periods(3, 1);
-  autopas::AutoPas_MPI_Cart_get(_communicator, 3, _decomposition.data(), periods.data(), _domainId.data());
-}
+  // have to cast away const because MPI requires it
+  autopas::AutoPas_MPI_Cart_get(_communicator, _dimensionCount, _decomposition.data(),
+                                const_cast<int *>(_periods.data()), _domainId.data());
 
-void RegularGridDecomposition::initializeLocalBox() { updateLocalBox(); }
-
-void RegularGridDecomposition::initializeNeighbourIds() {
-  for (int i = 0; i < 3; ++i) {
-    auto neighbourIndex = i * 2;
-    auto preceedingNeighbourId = _domainId;
-    preceedingNeighbourId[i] = (--preceedingNeighbourId[i] + _decomposition[i]) % _decomposition[i];
-    _neighbourDomainIndices[neighbourIndex] = DomainTools::convertIdToIndex(preceedingNeighbourId, _decomposition);
-
-    ++neighbourIndex;
-    auto succeedingNeighbourId = _domainId;
-    succeedingNeighbourId[i] = (++succeedingNeighbourId[i] + _decomposition[i]) % _decomposition[i];
-    _neighbourDomainIndices[neighbourIndex] = DomainTools::convertIdToIndex(succeedingNeighbourId, _decomposition);
+  // Create planar communicators used for diffuse load balancing.
+  for (int i = 0; i < _dimensionCount; ++i) {
+    if (_mpiCommunicationNeeded) {
+      const int key = _decomposition[(i + 1) % _dimensionCount] * _domainId[(i + 2) % _dimensionCount] +
+                      _domainId[(i + 1) % _dimensionCount];
+      autopas::AutoPas_MPI_Comm_split(_communicator, _domainId[i], key, &_planarCommunicators[i]);
+    } else {
+      _planarCommunicators[i] = AUTOPAS_MPI_COMM_WORLD;
+    }
   }
 }
 
-void RegularGridDecomposition::updateLocalBox() {
+void RegularGridDecomposition::initializeLocalBox() {
   for (int i = 0; i < 3; ++i) {
     double localBoxWidth = (_globalBoxMax[i] - _globalBoxMin[i]) / static_cast<double>(_decomposition[i]);
 
     _localBoxMin[i] = _domainId[i] * localBoxWidth + _globalBoxMin[i];
     _localBoxMax[i] = (_domainId[i] + 1) * localBoxWidth + _globalBoxMin[i];
+  }
+}
 
-    if (_domainId[i] == 0) {
-      _localBoxMin[i] = _globalBoxMin[i];
-    } else if (_domainId[i] == _decomposition[i] - 1) {
-      _localBoxMax[i] = _globalBoxMax[i];
-    }
+void RegularGridDecomposition::initializeNeighborIndices() {
+  for (int i = 0; i < 3; ++i) {
+    auto neighborIndex = i * 2;
+    auto preceedingNeighborId = _domainId;
+    preceedingNeighborId[i] = (--preceedingNeighborId[i] + _decomposition[i]) % _decomposition[i];
```

Comment:
Why add by `_decomposition[i]` just to modulo by `_decomposition[i]`?

Reply:
not 100% sure thb but it makes a difference if `preceedingNeighborId[i] == 0`. Then
```cpp
(--preceedingNeighborId[i] + _decomposition[i]) % _decomposition[i]  == _decomposition[i] - 1
```
but
```cpp
(--preceedingNeighborId[i]) % _decomposition[i]  ==  - 1
```
And _decomposition[i] is guaranteed to be `> 0`

So after all probably a safety thing :shrug:

Reply:
Okay yeah that makes sense